src/PFHub.hpp

```diff
@@ -109,14 +119,58 @@ class PFHub1aBase : public CabanaPFRunner<2> {
         vars.fft_inverse(0);
     }
 
+    // calculate the free energy (in between timesteps)
+    double free_energy() {
+        // copy over c for FFTing.  We do this instead of using vars[0] because FFTing and IFFTing could introduce
+        // additional floating point error and calculating free energy shouldn't change the future values of c
+        const auto concentration = vars[0];
+        const auto c_copy = vars[1];
+        node_parallel_for(
+            "free energy: copy c", KOKKOS_LAMBDA(const int i, const int j) {
+                c_copy(i, j, 0) = concentration(i, j, 0);
+                c_copy(i, j, 1) = concentration(i, j, 1);
+            });
+        // enter fourier space:
+        vars.fft_forward(1);
+        const auto c_hat = vars[1];
+        // calculate the gradients in Fourier space:
+        const auto gradient_x = vars[2], gradient_y = vars[3];
+        const auto kx = kx_view, ky = ky_view;
+        const double cell_size_1d = cell_size;
+        node_parallel_for(
+            "free energy: gradient", KOKKOS_LAMBDA(const int i, const int j) {
+                const cdouble c(c_hat(i, j, 0), c_hat(i, j, 1));
+                const cdouble grad_x = c * kx(i, j) / cell_size_1d;
+                gradient_x(i, j, 0) = grad_x.real();
+                gradient_x(i, j, 1) = grad_x.imag();
+                const cdouble grad_y = c * ky(i, j) / cell_size_1d;
+                gradient_y(i, j, 0) = grad_y.real();
+                gradient_y(i, j, 1) = grad_y.imag();
+            });
+        // return to real space:
+        vars.fft_inverse(2);
+        vars.fft_inverse(3);
+        // calculate the free energy:
+        const double dA = cell_size * cell_size;
+        const double KAPPA = _KAPPA, RHO = _RHO, C_ALPHA = _C_ALPHA, C_BETA = _C_BETA;
+        return node_parallel_reduce(
+            "free_energy", KOKKOS_LAMBDA(const int i, const int j, double& result) {
+                const double gradient_magnitude_squared =
+                    gradient_x(i, j, 0) * gradient_x(i, j, 0) + gradient_y(i, j, 0) * gradient_y(i, j, 0);
+                const double c = concentration(i, j, 0);
+                result += dA * (RHO * (c - C_ALPHA) * (c - C_ALPHA) * (c - C_BETA) * (c - C_BETA) +
+                                KAPPA * gradient_magnitude_squared / 2);
+            });
+    }
+
     // versions of PFHub1a must override this to give themselves a name:
     virtual std::string subproblem_name() const = 0;
     // save a copy of the c grid to a file
     void output_c() {
         std::stringstream s;
         s << subproblem_name() << "_N" << grid_points << "_DT" << std::fixed << std::setprecision(3) << std::scientific
           << dt;
-        vars.save(0, s.str(), 0);
+        vars.save(0, s.str(), get_timesteps_done(), get_time_done());
```

Comment:
Noting that I think this is where the output bug was fixed (in general, it's worth fixing these in separate PRs, but it's already done here)